https://github.com/protobuf-net/protobuf-net/issues/867#issue-1073597953

channell opened the pull request:
Added Serializers for F# Map / Set / List and test case

mgravell commented on src/protobuf-net.Test/protobuf-net.Test.csproj:
```diff
@@ -21,6 +21,10 @@
     <DefineConstants>FEAT_COMPILER;FX30</DefineConstants>
   </PropertyGroup>
   <ItemGroup>
+    <None Remove="C:\Users\steve\.nuget\packages\fsharp.core\6.0.1\contentFiles\any\netstandard2.1\FSharp.Core.xml" />
```
????

mgravell commented on src/protobuf-net.BuildTools/protobuf-net.BuildTools.csproj:
```diff
@@ -15,15 +15,16 @@
 
   <ItemGroup>
     <!-- build time deps -->
+    <PackageReference Include="FSharp.Core" Version="6.0.1" />
```
does protobuf-net.BuildTools[.Legacy] actually need this?

mgravell commented on src/protobuf-net.Core/Meta/TypeModel.cs:
```diff
@@ -132,15 +132,17 @@ internal static WireType GetWireType(TypeModel model, DataFormat format, Type ty
+        }
+
+        /// <summary>
+        /// Indicates whether a type is known to the model
```
the change here is fine as far as it goes, but if you look at the commit delta including whitespace: it's a doozy; honestly, it might be worth just reverting this one file and not worrying about this tweak

mgravell commented:
Overall: I like the idea, but I'm a little reluctant to add the dependency everywhere; that seems like something we can and should avoid - because right now:

![image](https://user-images.githubusercontent.com/17328/145090502-7c206b22-a254-4c55-8322-8f44ab280304.png)

![image](https://user-images.githubusercontent.com/17328/145090484-32fef2e5-b068-4a81-8644-aedd85865ec3.png)

I wonder if the way to go here is some kind of `protobuf-net.FSharp` package with some init method that registers the required additions (perhaps via some new API, even if we need to add an `[InternalsVisibleTo(...)]`). Thoughts?

channell commented:
> Overall: I like the idea, but I'm a little reluctant to add the dependency everywhere; that seems like something we can and should avoid - because right now:
> 
> ![image](https://user-images.githubusercontent.com/17328/145090502-7c206b22-a254-4c55-8322-8f44ab280304.png)
> 
> ![image](https://user-images.githubusercontent.com/17328/145090484-32fef2e5-b068-4a81-8644-aedd85865ec3.png)
> 
> I wonder if the way to go here is some kind of `protobuf-net.FSharp` package with some init method that registers the required additions (perhaps via some new API, even if we need to add an `[InternalsVisibleTo(...)]`). Thoughts?

That's a better idea..  I'd added the collection serializers because it is not possible to use surrogates for novel collection collection types because they need to be repeated.  Having an API would allow other collection types to plugged-in..

Happy to go with whatever you thinks best, but otherwise would add to the RuntimeModel in a similar way to SetSurrogate.
will look in the monrning

mgravell commented:
There probably isn't an existing API for this, but: we can change that :)

On Tue, 7 Dec 2021, 21:26 channell, ***@***.***> wrote:

> Overall: I like the idea, but I'm a little reluctant to add the dependency
> everywhere; that seems like something we can and should avoid - because
> right now:
>
> [image: image]
> <https://user-images.githubusercontent.com/17328/145090502-7c206b22-a254-4c55-8322-8f44ab280304.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/17328/145090484-32fef2e5-b068-4a81-8644-aedd85865ec3.png>
>
> I wonder if the way to go here is some kind of protobuf-net.FSharp
> package with some init method that registers the required additions
> (perhaps via some new API, even if we need to add an
> [InternalsVisibleTo(...)]). Thoughts?
>
> That's a better idea.. I'd added the collection serializers because it is
> not possible to use surrogates for novel collection collection types
> because they need to be repeated. Having an API would allow other
> collection types to plugged-in..
>
> Happy to go with whatever you thinks best, but otherwise would add to the
> RuntimeModel in a similar way to SetSurrogate.
> will look in the monrning
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/protobuf-net/protobuf-net/pull/868#issuecomment-988273960>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMBSSZO2JHHKTPB6OHTUPZ3X3ANCNFSM5JR3XXNA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

channell commented:
Hi Mark,

I took your advice and implemented F# support via an extension to the runtime model.  Can you take a look at my PR

Steve

Added API Extension for AddSerializer and of protbuf-net.FSharp by channell · Pull Request #875 · protobuf-net/protobuf-net (github.com)<https://github.com/protobuf-net/protobuf-net/pull/875>

From: Marc Gravell ***@***.***>
Sent: 07 December 2021 22:16
To: protobuf-net/protobuf-net ***@***.***>
Cc: channell ***@***.***>; Author ***@***.***>
Subject: Re: [protobuf-net/protobuf-net] https://github.com/protobuf-net/protobuf-net/issues/867#issue-1073597953 (PR #868)

There probably isn't an existing API for this, but: we can change that :)

On Tue, 7 Dec 2021, 21:26 channell, ***@***.***<mailto:***@***.***>> wrote:

> Overall: I like the idea, but I'm a little reluctant to add the dependency
> everywhere; that seems like something we can and should avoid - because
> right now:
>
> [image: image]
> <https://user-images.githubusercontent.com/17328/145090502-7c206b22-a254-4c55-8322-8f44ab280304.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/17328/145090484-32fef2e5-b068-4a81-8644-aedd85865ec3.png>
>
> I wonder if the way to go here is some kind of protobuf-net.FSharp
> package with some init method that registers the required additions
> (perhaps via some new API, even if we need to add an
> [InternalsVisibleTo(...)]). Thoughts?
>
> That's a better idea.. I'd added the collection serializers because it is
> not possible to use surrogates for novel collection collection types
> because they need to be repeated. Having an API would allow other
> collection types to plugged-in..
>
> Happy to go with whatever you thinks best, but otherwise would add to the
> RuntimeModel in a similar way to SetSurrogate.
> will look in the monrning
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/protobuf-net/protobuf-net/pull/868#issuecomment-988273960>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMBSSZO2JHHKTPB6OHTUPZ3X3ANCNFSM5JR3XXNA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/protobuf-net/protobuf-net/pull/868#issuecomment-988300417>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABNPQ7METN3ULJEKQGXY723UP2BQHANCNFSM5JR3XXNA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
